pkg/traceql/engine_metrics_compare.go

```diff
@@ -0,0 +1,439 @@
+		return
+	}
+}
+
+func (m *MetricsCompare) observe(span Span) {
+	// For performance, MetricsCompare doesn't use the Range/StepAggregator abstractions.
+	// This lets us:
+	// * Include the same attribute value in multiple series. This doesn't fit within
+	//   the existing by() grouping or even the potential byeach() (which was in this branch and then deleted)
+	// * Avoid reading the span start time twice, once for the selection window filter, and
+	//   then again instead of StepAggregator.
+	// TODO - It would be nice to use those abstractions, area for future improvement
+	st := span.StartTimeUnixNanos()
+	i := IntervalOf(st, m.qstart, m.qend, m.qstep)
+
+	// Determine if this span is inside the selection
+	isSelection := StaticFalse
+	if m.start > 0 && m.end > 0 {
+		// Timestamp filtering
+		if st >= uint64(m.start) && st < uint64(m.end) {
+			isSelection, _ = m.f.Expression.execute(span)
+		}
+	} else {
+		// No timestamp filtering
+		isSelection, _ = m.f.Expression.execute(span)
+	}
+
+	// Choose destination buffers
+	dest := m.baselines
+	destTotals := m.baselineTotals
+	if isSelection == StaticTrue {
+		dest = m.selections
+		destTotals = m.selectionTotals
+	}
+
+	// Increment values for all attributes of this span
+	span.AllAttributesFunc(func(a Attribute, v Static) {
+		// These attributes get pulled back by select all but we never
+		// group by them because I say so.
+		// TODO - can we check type instead?
+		switch a {
+		case IntrinsicSpanStartTimeAttribute, IntrinsicDurationAttribute:
+			return
+		}
+
+		values, ok := dest[a]
+		if !ok {
+			values = make(map[Static][]float64, m.len)
+			dest[a] = values
+		}
+
+		counts, ok := values[v]
+		if !ok {
+			counts = make([]float64, m.len)
+			values[v] = counts
+		}
+		counts[i]++
+
+		// TODO - It's probably faster to aggregate these at the end
+		// instead of incrementing in the hotpath twice
+		totals, ok := destTotals[a]
+		if !ok {
+			totals = make([]float64, m.len)
+			destTotals[a] = totals
+		}
+		totals[i]++
+	})
+}
+
+func (m *MetricsCompare) observeSeries(ss []*tempopb.TimeSeries) {
+	m.seriesAgg.Combine(ss)
+}
+
+func (m *MetricsCompare) result() SeriesSet {
+	// In the other modes return these results
+	if m.seriesAgg != nil {
+		return m.seriesAgg.Results()
+	}
+
+	var (
+		top   = topN[Static]{}
+		ss    = make(SeriesSet)
+		erred = make(map[Attribute]struct{})
+	)
+
+	add := func(ls Labels, counts []float64) {
+		ss[ls.String()] = TimeSeries{
+			Labels: ls,
+			Values: counts,
+		}
+	}
+
+	addValues := func(prefix Label, data map[Attribute]map[Static][]float64) {
+		for a, values := range data {
+			// Compute topN values for this attribute
+			top.reset()
+			for v, counts := range values {
+				top.add(v, counts)
+			}
+
+			top.get(m.topN, func(v Static) {
+				add(Labels{
+					prefix,
+					{Name: a.String(), Value: v},
+				}, values[v])
+			})
+
+			if len(values) > m.topN {
+				erred[a] = struct{}{}
+			}
```

Comment:
Shouldn't it collect all possible values and return the top N when merging the results instead?

Reply:
Solved async: doing this for performance reasons.

Reply:
Due to cardinality, it's not feasible to do a true topN on the frontend.  Each querier job could be sending 10-100K time series, and the frontend could be receiving millions of series. Therefore it computes a local topN per job first. This loses some precision on the edges but is a worthwhile tradeoff.